src/buffer/out/ut_textbuffer/ReflowTests.cpp

```diff
@@ -0,0 +1,856 @@
+#include "../../types/inc/Utf16Parser.hpp"
+#include "../../types/inc/GlyphWidth.hpp"
+
+#include <IDataSource.h>
+
+template<>
+class WEX::TestExecution::VerifyOutputTraits<wchar_t>
+{
+public:
+    static WEX::Common::NoThrowString ToString(const wchar_t& wch)
+    {
+        return WEX::Common::NoThrowString().Format(L"'%c'", wch);
+    }
+};
+
+using namespace WEX::Common;
+using namespace WEX::Logging;
+using namespace WEX::TestExecution;
+
+namespace
+{
+    struct TestRow
+    {
+        std::wstring_view text;
+        bool wrap;
+    };
+
+    struct TestBuffer
+    {
+        COORD size;
+        std::vector<TestRow> rows;
+        COORD cursor;
+    };
+
+    struct TestCase
+    {
+        std::wstring_view name;
+        std::vector<TestBuffer> buffers;
+    };
+
+    static constexpr auto true_due_to_exact_wrap_bug{ true };
+
+    static const TestCase testCases[] = {
+        TestCase{
+            L"No reflow required",
+            {
+                TestBuffer{
+                    { 6, 5 },
+                    {
+                        { L"AB    ", false },
+                        { L"$     ", false },
+                        { L"CD    ", false },
+                        { L"EFG   ", false },
+                        { L"      ", false },
+                    },
+                    { 0, 1 } // cursor on $
+                },
+                TestBuffer{
+                    { 5, 5 }, // reduce width by 1
+                    {
+                        { L"AB   ", false },
+                        { L"$    ", false },
+                        { L"CD   ", false },
+                        { L"EFG  ", false },
+                        { L"     ", false },
+                    },
+                    { 0, 1 } // cursor on $
+                },
+                TestBuffer{
+                    { 4, 5 },
+                    {
+                        { L"AB  ", false },
+                        { L"$   ", false },
+                        { L"CD  ", false },
+                        { L"EFG ", false },
+                        { L"    ", false },
+                    },
+                    { 0, 1 } // cursor on $
+                },
+            },
+        },
+        TestCase{
+            L"SBCS, cursor remains in buffer, no circling, no original wrap",
+            {
+                TestBuffer{
+                    { 6, 5 },
+                    {
+                        { L"ABCDEF", false },
+                        { L"$     ", false },
+                        { L"      ", false },
+                        { L"      ", false },
+                        { L"      ", false },
+                    },
+                    { 0, 1 } // cursor on $
+                },
+                TestBuffer{
+                    { 5, 5 }, // reduce width by 1
+                    {
+                        { L"ABCDE", true },
+                        { L"F$   ", false }, // [BUG] EXACT WRAP. $ should be alone on next line.

```

Comment:
Should these [BUG] tags be filed and linked appropriately?

Reply:
I'm probably not gonna file 20 new bugs for "reflow is broken", which is a well-known statement.

I should have used a less controversial word, like "[KNOWN BADNESS]"